### DISCUSSION TITLE
MNK 7.0

### DISCUSSION THREAD

**Discussion by Evolutious** [PrincessRTFM/XIVComboPlugin#106] [General]
Hey @PrincessRTFM! Long time!

I couldn't help but notice your comment about "fuck monk" ROFL!!

So yeah, the MNK gauge is the problem right now. It got changed a good bit by SE and there was a entirely new component added to it  (the new balls). I was able to get the Beast Chakras corrected by adjusting which action you want to use based on which Beast Chakra is missing, but the Nadi do not align at all.

So, current state for MNK is it needs lots of JobGauge updates to Dalamud before anyone is going to be able to get it working very well (not to mention that the gauges are getting another tweak in 7.01). I have resorted to Frankenstein and using a combination of XIVCombo plus AutoHotLKey PixlelScan to make things work, decently.

Once Dalamud gets the MNK job gauge updates, I will be happy to look and see if I can clean up your combo.

**Comment by Evolutious** [comment]
Not sure how you would want to incorporate this into your code, but this is what I found looking at how the DelvUI guys fixed the MNK gauge issues. I got it working in my code by adding this to the internal class combo:

```csharp
        // New Fury Gauge
        ushort fury = *((ushort*)(new IntPtr(gauge.Address) + 0xC));
        int opoOpoFury = fury & 1;
        int raptorFury = ((fury >> 2) & 3);
        int coeurlFury = ((fury >> 4) & 3);

        // Adjusted Nadi Gauge
        ushort nadi = *((ushort*)(new IntPtr(gauge.Address) + 0xD));
```

The challenge with this is it has to be placed inside an ```unsafe``` class. I just changed the ```internal class``` combos to being ```unsafe internal class```. You may want to move this to another area in your code that you can then just reference in the combo class.

A few notes about the above as it relates to the MNK gauges:
```opoOpoFury``` returns the number of the new Opo-Opo fury stacks (0-1)
```raptorFury``` returns the number of the new Raptor fury stacks (0-2)
```coeurlFury``` returns the number of the new Coeurl fury stacks (0-3)
```nadi``` returns 0-3 depending on your Nadi state
- 0 = Neither Lunar Nadi nor Solar Nadi
- 1 = Lunar Nadi Only
- 2 = Solar Nadi Only
- 3 = Both Lunar Nadi and Solar Nadi

SE also changed which Beast Chakra you are given based on the action you use:
- Bootshine/Dragon Kick/Leaping Opo = COEURL
- True Strike/Twin Snakes/Rising Raptor = OPOOPO
- Snap Punch/Demolish/Pouncing Coeurl = RAPTOR

If you want to think about how you want to incorporate the code above to add the new Fury gauge and the adjusted Nadi gauge components, I can help make the needed changes in your MNK combo to get it working again.

An example of how this plays out in code:
```csharp
        // Master's Gauge
        if (SelfHasEffect(MNK.Buffs.PerfectBalance))
        {
            // Solar
            if (level >= MNK.Levels.EnhancedPerfectBalance && nadi < 2)
            {
                if (!gauge.BeastChakra.Contains(BeastChakra.COEURL))
                {
                    return (opoOpoFury > 0)
                        ? OriginalHook(MNK.Bootshine)
                        : MNK.DragonKick;
                }
                if (!gauge.BeastChakra.Contains(BeastChakra.OPOOPO))
                {
                    return (raptorFury > 0)
                        ? OriginalHook(MNK.TrueStrike)
                        : MNK.TwinSnakes;
                }
                if (!gauge.BeastChakra.Contains(BeastChakra.RAPTOR))
                {
                    return (coeurlFury > 0)
                        ? OriginalHook(MNK.SnapPunch)
                        : MNK.Demolish;
                }
            }
            // Lunar.  Also used if we have both Nadi as Tornado Kick/Phantom Rush isn't picky, or under 60.
            return (opoOpoFury > 0)
                ? OriginalHook(MNK.Bootshine)
                : MNK.DragonKick;
        }
```

**Reply by VariableVixen** [reply]
I haven't written any of the existing MNK stuff, and frankly, I don't think I'm going to. It's the one job that I just _can't_ wrap my brain around. But if you want to wait for dalamud to update the job gauge stuff and write it yourself, I'll give it a lookover and merge it as long as nothing jumps out!
